provider/pihole/clientV6.go

```diff
@@ -0,0 +1,376 @@
+			CnameRecords []string `json:"cnameRecords"`
+		} `json:"dns"`
+	} `json:"config"`
+	Took float64 `json:"took"`
+}
+
+func (p *piholeClientV6) apply(ctx context.Context, action string, ep *endpoint.Endpoint) error {
+	if !p.cfg.DomainFilter.Match(ep.DNSName) {
+		log.Debugf("Skipping %s %s that does not match domain filter", action, ep.DNSName)
+		return nil
+	}
+	apiUrl, err := p.urlForRecordType(ep.RecordType)
+	if err != nil {
+		log.Warnf("Skipping unsupported endpoint %s %s %v", ep.DNSName, ep.RecordType, ep.Targets)
+		return nil
+	}
+
+	if p.cfg.DryRun {
+		log.Infof("DRY RUN: %s %s IN %s -> %s", action, ep.DNSName, ep.RecordType, ep.Targets[0])
+		return nil
+	}
+
+	log.Infof("%s %s IN %s -> %s", action, ep.DNSName, ep.RecordType, ep.Targets[0])
+
+	// Get the current record
+	if strings.Contains(ep.DNSName, "*") {
+		return provider.NewSoftError(errors.New("UNSUPPORTED: Pihole DNS names cannot return wildcard"))
+	}
+
+	switch ep.RecordType {
+	case endpoint.RecordTypeA, endpoint.RecordTypeAAAA:
+		apiUrl = fmt.Sprintf("%s/%s", apiUrl, url.PathEscape(fmt.Sprintf("%s %s", ep.Targets, ep.DNSName)))
+	case endpoint.RecordTypeCNAME:
+		if ep.RecordTTL.IsConfigured() {
+			apiUrl = fmt.Sprintf("%s/%s", apiUrl, url.PathEscape(fmt.Sprintf("%s,%s,%d", ep.DNSName, ep.Targets, ep.RecordTTL)))
+		} else {
+			apiUrl = fmt.Sprintf("%s/%s", apiUrl, url.PathEscape(fmt.Sprintf("%s,%s", ep.DNSName, ep.Targets)))
+		}
+	}
+
+	req, err := http.NewRequestWithContext(ctx, action, apiUrl, nil)
+	if err != nil {
+		return err
+	}
+
+	_, err = p.do(req)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *piholeClientV6) retrieveNewToken(ctx context.Context) error {
+	if p.cfg.Password == "" {
+		return nil
+	}
+
+	apiUrl := fmt.Sprintf("%s/api/auth", p.cfg.Server)
+	log.Debugf("Fetching new token from %s", apiUrl)
+
+	// Define the JSON payload
+	jsonData := []byte(`{"password":"` + p.cfg.Password + `"}`)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, apiUrl, bytes.NewBuffer(jsonData))
+	if err != nil {
+		return err
+	}
+
+	jRes, err := p.do(req)
+	if err != nil {
+		return err
+	}
+
+	// Parse JSON response
+	var apiResponse ApiAuthResponse
+	err = json.Unmarshal(jRes, &apiResponse)
+	if err != nil {
+		fmt.Println("Error reading response:", err)
+	} else {
+		// Set the token
+		if apiResponse.Session.SID != "" {
+			p.token = apiResponse.Session.SID
+		}
+	}
+	return err
+}
+
+func (p *piholeClientV6) checkTokenValidity(ctx context.Context) (bool, error) {
+	if p.token == "" {
+		return false, nil
+	}
+
+	apiUrl := fmt.Sprintf("%s/api/auth", p.cfg.Server)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, apiUrl, nil)
+	if err != nil {
+		return false, nil
+	}
+
+	jRes, err := p.do(req)
+	if err != nil {
+		return false, nil
+	}
+
+	// Parse JSON response
+	var apiResponse ApiAuthResponse
+	err = json.Unmarshal(jRes, &apiResponse)
+	if err != nil {
+		fmt.Println("Error reading response:", err)
```

Comment:
if err := json.Unmarshal(jRes, &apiError); err != nil {
return nil, fmt.Errorf("failed to unmarshal error response: %w", err)
}